File: backend/admin/utils/session_manager.py
"""
Session Manager Utility

Manages admin user sessions with 8-hour timeout.
"""

import time
from typing import Dict, Any, Optional
import boto3
from datetime import datetime, timedelta

# AWS clients
dynamodb = boto3.resource("dynamodb")


class SessionManager:
    """Manages admin user sessions"""
    
    def __init__(self, table_name: str = "SanaathanaAalayaCharithra-AdminSessions"):
        self.table = dynamodb.Table(table_name)
        self.session_timeout_hours = 8
    
    def create_session(self, user_id: str, token_jti: str) -> str:
        """
        Create new session
        
        Args:
            user_id: User identifier
            token_jti: JWT token ID
            
        Returns:
            Session ID
        """
        try:
            current_time = int(time.time())
            expires_at = current_time + (self.session_timeout_hours * 3600)
            
            session_id = f"{user_id}:{token_jti}"
            
            self.table.put_item(
                Item={
                    "sessionId": session_id,
                    "userId": user_id,
                    "tokenJti": token_jti,
                    "createdAt": current_time,
                    "expiresAt": expires_at,
                    "lastActivity": current_time,
                    "ttl": expires_at + 3600,  # Auto-delete 1 hour after expiry
                }
            )
            
            return session_id
            
        except Exception as e:
            print(f"Error creating session: {str(e)}")
            raise
    
    def validate_session(self, user_id: str, token_jti: str) -> bool:
        """
        Validate session is active and not expired
        
        Args:
            user_id: User identifier
            token_jti: JWT token ID
            
        Returns:
            True if session is valid, False otherwise
        """
        try:
            session_id = f"{user_id}:{token_jti}"
            
            response = self.table.get_item(
                Key={"sessionId": session_id}
            )
            
            if "Item" not in response:
                return False
            
            session = response["Item"]
            current_time = int(time.time())
            
            # Check if session expired
            if session["expiresAt"] < current_time:
                return False
            
            # Update last activity
            self.table.update_item(
                Key={"sessionId": session_id},
                UpdateExpression="SET lastActivity = :activity",
                ExpressionAttributeValues={
                    ":activity": current_time,
                },
            )
            
            return True
            
        except Exception as e:
            print(f"Error validating session: {str(e)}")
            return False
    
    def terminate_session(self, user_id: str, token_jti: str) -> None:
        """
        Terminate session
        
        Args:
            user_id: User identifier
            token_jti: JWT token ID
        """
        try:
            session_id = f"{user_id}:{token_jti}"
            self.table.delete_item(Key={"sessionId": session_id})
        except Exception as e:
            print(f"Error terminating session: {str(e)}")
    
    def terminate_all_user_sessions(self, user_id: str) -> int:
        """
        Terminate all sessions for a user
        
        Args:
            user_id: User identifier
            
        Returns:
            Number of sessions terminated
        """
        try:
            # Query all sessions for user
            response = self.table.query(
                IndexName="UserIdIndex",
                KeyConditionExpression="userId = :uid",
                ExpressionAttributeValues={":uid": user_id},
            )
            
            sessions = response.get("Items", [])
            count = 0
            
            for session in sessions:
                self.table.delete_item(
                    Key={"sessionId": session["sessionId"]}
                )
                count += 1
            
            return count
            
        except Exception as e:
            print(f"Error terminating user sessions: {str(e)}")
            return 0

File: backend/src/auth/middleware/session-manager.ts
/**
 * Session Manager Utility
 * 
 * Manages admin user sessions with 8-hour timeout.
 */

import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, GetCommand, PutCommand, UpdateCommand, DeleteCommand, QueryCommand } from '@aws-sdk/lib-dynamodb';

const dynamoClient = new DynamoDBClient({ region: process.env.AWS_REGION || 'us-east-1' });
const docClient = DynamoDBDocumentClient.from(dynamoClient);

export interface SessionEntry {
  sessionId: string;
  userId: string;
  tokenJti: string;
  createdAt: number;
  expiresAt: number;
  lastActivity: number;
  ttl: number;
}

export class SessionManager {
  private tableName: string;
  private sessionTimeoutHours: number;

  constructor(tableName: string = 'SanaathanaAalayaCharithra-AdminSessions') {
    this.tableName = tableName;
    this.sessionTimeoutHours = 8;
  }

  /**
   * Create new session
   */
  async createSession(userId: string, tokenJti: string): Promise<string> {
    try {
      const currentTime = Math.floor(Date.now() / 1000);
      const expiresAt = currentTime + (this.sessionTimeoutHours * 3600);
      const sessionId = `${userId}:${tokenJti}`;

      const putCommand = new PutCommand({
        TableName: this.tableName,
        Item: {
          sessionId,
          userId,
          tokenJti,
          createdAt: currentTime,
          expiresAt,
          lastActivity: currentTime,
          ttl: expiresAt + 3600, // Auto-delete 1 hour after expiry
        },
      });

      await docClient.send(putCommand);
      return sessionId;

    } catch (error) {
      console.error(`Error creating session: ${error}`);
      throw error;
    }
  }

  /**
   * Validate session is active and not expired
   */
  async validateSession(userId: string, tokenJti: string): Promise<boolean> {
    try {
      const sessionId = `${userId}:${tokenJti}`;

      const getCommand = new GetCommand({
        TableName: this.tableName,
        Key: { sessionId },
      });

      const response = await docClient.send(getCommand);

      if (!response.Item) {
        return false;
      }

      const session = response.Item as SessionEntry;
      const currentTime = Math.floor(Date.now() / 1000);

      // Check if session expired
      if (session.expiresAt < currentTime) {
        return false;
      }

      // Update last activity
      const updateCommand = new UpdateCommand({
        TableName: this.tableName,
        Key: { sessionId },
        UpdateExpression: 'SET lastActivity = :activity',
        ExpressionAttributeValues: {
          ':activity': currentTime,
        },
      });

      await docClient.send(updateCommand);
      return true;

    } catch (error) {
      console.error(`Error validating session: ${error}`);
      return false;
    }
  }

  /**
   * Terminate session
   */
  async terminateSession(userId: string, tokenJti: string): Promise<void> {
    try {
      const sessionId = `${userId}:${tokenJti}`;

      const deleteCommand = new DeleteCommand({
        TableName: this.tableName,
        Key: { sessionId },
      });

      await docClient.send(deleteCommand);

    } catch (error) {
      console.error(`Error terminating session: ${error}`);
    }
  }

  /**
   * Terminate all sessions for a user
   */
  async terminateAllUserSessions(userId: string): Promise<number> {
    try {
      // Query all sessions for user
      const queryCommand = new QueryCommand({
        TableName: this.tableName,
        IndexName: 'UserIdIndex',
        KeyConditionExpression: 'userId = :uid',
        ExpressionAttributeValues: { ':uid': userId },
      });

      const response = await docClient.send(queryCommand);
      const sessions = response.Items || [];
      let count = 0;

      for (const session of sessions) {
        const deleteCommand = new DeleteCommand({
          TableName: this.tableName,
          Key: { sessionId: session.sessionId },
        });

        await docClient.send(deleteCommand);
        count++;
      }

      return count;

    } catch (error) {
      console.error(`Error terminating user sessions: ${error}`);
      return 0;
    }
  }

  /**
   * Get session details (for testing)
   */
  async getSession(userId: string, tokenJti: string): Promise<SessionEntry | null> {
    try {
      const sessionId = `${userId}:${tokenJti}`;

      const getCommand = new GetCommand({
        TableName: this.tableName,
        Key: { sessionId },
      });

      const response = await docClient.send(getCommand);
      return response.Item as SessionEntry || null;

    } catch (error) {
      console.error(`Error getting session: ${error}`);
      return null;
    }
  }

  /**
   * Check if session is expired (for testing)
   */
  async isSessionExpired(userId: string, tokenJti: string): Promise<boolean> {
    try {
      const session = await this.getSession(userId, tokenJti);
      if (!session) {
        return true;
      }

      const currentTime = Math.floor(Date.now() / 1000);
      return session.expiresAt < currentTime;

    } catch (error) {
      console.error(`Error checking session expiry: ${error}`);
      return true;
    }
  }
}